### PR TITLE
Refactor the zone resource

### PR DESCRIPTION
**Pull request by eherot**
Sorry, the branch is misnamed. Also there are some other things in here that had to happen in concert.

This does not include the "cluster" functionality. That comes next.

### CODE REVIEW

File: libraries/ec2_dns_server.rb [hunk 1]
```diff
@@ -139,43 +167,23 @@ def static_record_nodenames(static_records = {})
...
-
   def non_public_interfaces(server)
+    unless zone_options['avoid_subnets'] == []
+      Chef::Log.info('Avoiding these subnets: ' +
+        zone_options['avoid_subnets'].join(','))
+    end
+
     server.network_interfaces.reject do |ni|
-      @avoid_subnets.include?(ni['subnetId']) || ni == {}
+      zone_options['avoid_subnets'].include?(ni['subnetId']) || ni == {}
```
Comment (jeffbyrnes) [inline libraries/ec2_dns_server.rb]: A nit, but maybe it’d be better to use the more common shorthand of `nic` for the variable here? On the first read, I thought `ni` was a typo.

File: spec/default_spec.rb [hunk 2]
```diff
@@ -1,160 +1,577 @@
 require 'spec_helper'
+require 'byebug'
```
Comment (jeffbyrnes) [inline spec/default_spec.rb]: I’m thinking this shouldn’t still be here.

File: spec/default_spec.rb [hunk 3]
```diff
@@ -1,160 +1,577 @@
...
+          fog_conn.attach_network_interface(
+            public_interface_id,
+            server.id,
+            '1'
+          )
+        end
+        # A public IP is expected here because, lacking any non-excluded IPs to list,
+        # the server_obj_ip falls back to server.private_ip_address which ends up
+        # being the public IP because Fog mocking doesn't understand that
+        # 'avoid_subnets' refers to public interfaces.
```
Comment (jeffbyrnes) [inline spec/default_spec.rb]: Super helpful comment here; that‘s a nasty bit of nuance to Fog mocking.

File: spec/default_spec.rb [hunk 4]
```diff
@@ -1,160 +1,577 @@
...
+  #   context 'target server does not exist' do
+  #     it 'raises an exception' do
+  #       expect { helpers.send(:server_ip_by_hostname, 'some-node') }
+  #         .to raise_error(
+  #           RuntimeError,
+  #           "Can't locate server with name some-node"
+  #         )
+  #     end
+  #   end
+  # end
```
Comment (jeffbyrnes) [inline spec/default_spec.rb]: Why are these all commented out?

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -1,13 +1,18 @@
 require 'chefspec'
 require 'chefspec/berkshelf'
+require 'hashie'
 require 'fog'
+require 'ipaddress'
 # require 'rspec/mocks'
 
-require_relative '../libraries/Ec2DnsServer.rb'
+require_relative '../libraries/ec2_dns_server.rb'
```
Comment (jeffbyrnes) [inline spec/spec_helper.rb]: :clap:

**Comment by jeffbyrnes** [comment]
@eherot I’ll give it another read tomorrow AM if you want, but my first, somewhat quick, pass, it looks pretty solid. Definite improvements here, much more broken down to individual parts.

The extensive RSpec tests on the library are awesome.

File: spec/default_spec.rb (hunk 4 above)
Comment (eherot) [inline spec/default_spec.rb]: Oops. Meant to take that out. I didn't want to delete that code without committing it somewhere.

File: libraries/ec2_dns_server.rb [hunk 6]
```diff
@@ -38,91 +41,116 @@ def self.vpc_default_dns(vpc_net)
...
 
   def connection
     @connection ||= begin
       require 'fog'
 
-      aws_keys = Chef::EncryptedDataBagItem.load(
-        'secrets',
-        'aws_credentials'
-      )[@node['ec2dnsserver']['aws_api_user']]
+      Fog.mock! if mocking?
```
Comment (jeffbyrnes) [inline libraries/ec2_dns_server.rb]: @eherot seems like here, you could include some of what I’d done on my cleanup/test branch to set up a well-mocked env.

**Comment by jeffbyrnes** [comment]
@eherot gave this another pass. Lots happening, but really strong improvements, methinks. :+1:

**Comment by jeffbyrnes** [comment]
@eherot I’m thinking this should probably merge in & get out the door. Grown quite stale.

**Comment by eherot** [comment]
We should probably refactor this as a custom resource before actually merging it.

**Comment by eherot** [comment]
Thankfully although it is indeed old, the cookbook hasn't changed much since it was originally released.

**Comment by jeffbyrnes** [comment]
Agreed, though I’d argue we can get this merged in, and handle the custom resource refactor separately/additionally, since at this point, it’s basically “new work” to tackle that (owing to how long its been since either of us was working on this).

I’ll put a card in the inbox to handle refactoring this into a Chef 12.5 custom resource, since this cookbook is pretty critical & should be maintained.

**Comment by jeffbyrnes** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;glQXnIrb&#x2F;977-refactor-ec2dnsserver-zone-into-custom-resource"> Refactor ec2dnsserver_zone into custom resource</a>

File: libraries/ec2_dns_server.rb (hunk 1 above)
Comment (jeffbyrnes) [inline libraries/ec2_dns_server.rb]: Nevermind this nit; I get it now.

Comment (eherot) [inline libraries/ec2_dns_server.rb]: Sorry, didn't mean to ignore this earlier: It was a question of keeping with an industry standard term for Network Interface Cards vs. a coding standard of using acronyms for longer object names when working with them in a loop.

Comment (eherot) [inline libraries/ec2_dns_server.rb]: (In case that wasn't clear)

Comment (jeffbyrnes) [inline libraries/ec2_dns_server.rb]: Yep, that became clear as I worked with this code more. Thanks for confirming!

**Comment by jeffbyrnes** [comment]
@eherot this has been rebased again, against the current master branch.

At the moment, we see 4 tests failing:

```
25 examples, 4 failures

Failed examples:

rspec /tmp/verifier/suites/serverspec/default_spec.rb:36 # Zone Data Command "dig SOA priv.yourdomain.local @localhost" stdout should match /status: NOERROR/
rspec /tmp/verifier/suites/serverspec/default_spec.rb:37 # Zone Data Command "dig SOA priv.yourdomain.local @localhost" stdout should match /hostmaster\.yourdomain\.local\./
rspec /tmp/verifier/suites/serverspec/default_spec.rb:89 # Overrides Command "dig +short test-storm.priv.yourdomain.local @localhost" stdout should match /prod-dns-1c\.priv\.yourdomain\.local\./
rspec /tmp/verifier/suites/serverspec/default_spec.rb:101 # Company specific overrides Command "dig +short test-cookbook-host.yourdomain.local @localhost" stdout should match "10.0.6.150"
```

I spun up the master branch & compared some things:

```
ubuntu@ip-10-178-176-124:~$ cat /etc/bind/db.test-cookbook-host.yourdomain.local
$TTL 300
$ORIGIN test-cookbook-host.yourdomain.local.
@ 300 IN SOA default-ubuntu-1404.vagrantup.com. hostmaster.yourdomain.local. (
  1459179858
  3600
  600
  86400
  300
)

 NS default-ubuntu-1404.vagrantup.com.

 IN A
```

vs the working:

```
vagrant@default-ubuntu-1404:~$ cat /etc/bind/db.test-cookbook-host.yourdomain.local
$TTL 300
$ORIGIN test-cookbook-host.yourdomain.local.
@ 300 IN SOA default-ubuntu-1404.vagrantup.com. hostmaster.yourdomain.local. (
  1459180477
  3600
  600
  86400
  300
)

 NS default-ubuntu-1404.vagrantup.com.

 IN A 10.0.6.150
```

**Comment by eherot** [comment]
This implies two things to me:
- The check on the stub zone method to make sure the RR value is not blank is not working
- The recipe search is not finding the test node

**Comment by eherot** [comment]
I tested this branch using the `et_ec2dnsserver` cookbook which includes working versions of most of the same tests and got all green. I'm going to merge this (finally) and we can worry about fixing or consolidating the tests later.

**Comment by jeffbyrnes** [comment]
@eherot awesome! Long overdue, dang.

**Pull request merged**